Migration to discord.py 2.0

Jackenmen opened the pull request:
### Description of the changes

This is the work on migrating Red to discord.py 2.0.

If you want to test our work on migration to discord.py 2.0, you can test these changes by installing this PR's branch in a virtual environment with the command:
```
python -m pip install -U --force-reinstall git+https://github.com/Cog-Creators/Red-DiscordBot@V3/dpy2_unstable#egg=Red-DiscordBot
```
Be sure to report any issues you run into, this is a huge changeset so it's possible that I missed something.

Things that still need doing (but might be done in separate PR later):
- Quite a bit of user-facing docstrings refer to the current channel
  when they also mean thread (should we fix this?).
  - Developer-facing docs do too but I don't think that's an issue.
- Perhaps something we missed?

### Have the changes in this PR been tested?

No (Partially yes, but I have not tested all the changes that have been made here)

### Information for reviewers

The code review is best done by going through individual commits (there's a nice next button when you are in a commit view on a PR so it's quite convenient) as they are well-scoped.

As for functional review... You have to test on the whole changeset but you can still leverage the commit separation so that you test a single thing at the time though if you think it's going to be easier to just go through the whole files changed list, that seems fine to me too.

Jackenmen committed:
Temporarily set d.py to use latest git revision

Jackenmen committed:
Remove `bot` param to Client.start

Jackenmen committed:
Switch to aware datetimes

A lot of this is removing `.replace(...)` which while not technically
needed, simplifies the code base. There's only a few changes that are
actually necessary here.

Jackenmen committed:
Update to work with new Asset design

Jackenmen committed:
[threads] Update Filter cog to properly handle threads

- `[p]filter channel list` in a threads sends list for parent channel
- Checking for filter hits for a message in a thread checks its parent
  channel's word list. There's no separate word list for threads.

Jackenmen committed:
[threads] Update Red.message_eligible_as_command to support threads

Jackenmen committed:
[threads] Update type hints and documentation to reflect thread support

- Documented that `{channel}` in CCs might be a thread
- Allowed (documented) usage of threads with `Config.channel()`
    - Separate thread scope is still in the picture though
      if it were to be done, it's going to be in separate in PR
- GuildContext.channel might be Thread

Jackenmen committed:
Update for in-place edits

Jackenmen committed:
[threads] Disallow usage of threads in some of the commands

- announceset channel
- filter channel clear
- filter channel add
- filter channel remove
- GlobalUniqueObjectFinder converter
    - permissions addglobalrule
    - permissions removeglobalrule
    - permissions removeserverrule
    - Permissions cog does not perform any validation for IDs
      when setting through YAML so that has not been touched
- streamalert twitch/youtube/picarto
- embedset channel
- set ownernotifications adddestination

Jackenmen committed:
[threads] Properly handle invocation of [p](un)mutechannel in threads

Usage of a (un)mutechannel will mute/unmute user in the parent channel
if it's invoked in a thread.

Jackenmen committed:
[threads] Handle threads in Red's permissions system (Requires)

- Made permissions system apply rules of (only) parent in threads

Jackenmen committed:
[threads] Support threads in Audio cog

- Handle threads being notify channels
- Update type hint for `is_query_allowed()`

Jackenmen committed:
[threads] Update embed_requested to support threads

- Threads don't have their own embed settings and inherit from parent

Jackenmen committed:
[threads] Update core ModLog API to support threads

- Added proper support for passing `Thread` to `channel`
  when creating/editing case
- Added `parent_channel_id` attribute to Modlog API's Case
    - Added `parent_channel` property that tries to get parent channel
- Updated case's content to show both thread and parent information

Jackenmen committed:
Use less costy channel check in customcom's on_message_without_command

This isn't needed for d.py 2.0 but whatever...

Jackenmen committed:
Embed's bool changed behavior, I'm hoping it doesn't affect us

Jackenmen committed:
Address User.permissions_in() removal

Jackenmen committed:
Swap VerificationLevel.extreme with VerificationLevel.highest

Jackenmen committed:
Change to keyword-only parameters

Jackenmen committed:
Change of `Guild.vanity_invite()` return type

Jackenmen committed:
avatar -> display_avatar

Jackenmen committed:
Fix metaclass shenanigans with Converter

Jackenmen committed:
Update Red.add_cog() to be inline with `dpy_commands.Bot.add_cog()`

This means adding `override` keyword-only parameter and causing
small breakage by swapping RuntimeError with discord.ClientException.

Jackenmen committed:
Address all DEP-WARNs

Jackenmen committed:
Remove Context.clean_prefix and use upstream implementation instead

Jackenmen committed:
Remove commands.Literal and use upstream implementation instead

Honestly, this was a rather bad implementation anyway...

Breaking but actually not really - it was provisional.

Jackenmen committed:
Update Command.callback's setter

Support for functools.partial is now built into d.py

Jackenmen committed:
Add new perms in HUMANIZED_PERM mapping (some from d.py 1.7 it seems)

BTW, that should really be in core instead of what we have now...

Jackenmen committed:
Remove the part of do_conversion that has not worked for a long while

Jackenmen committed:
Stop wrapping BadArgument in ConversionFailure

This is breaking but it's best to resolve it like this.

The functionality of ConversionFailure can be replicated with
Context.current_parameter and Context.current_argument.

Jackenmen committed:
Add custom errors for int and float converters

Jackenmen committed:
Remove Command.__call__ as it's now implemented in d.py

Jackenmen committed:
Get rid of _dpy_reimplements

These were reimplemented for the purpose of typing
so it is no longer needed now that d.py is type hinted.

Jackenmen committed:
Add return to Red.remove_cog

Jackenmen committed:
Ensure we don't delete messages that differ only by used sticker

Jackenmen committed:
discord.InvalidArgument->ValueError

Jackenmen committed:
Move from raw <t:...> syntax to discord.utils.format_dt()

Jackenmen committed:
Address AsyncIter removal

Jackenmen committed:
Swap to pos-only for params that are pos-only in upstream

Jackenmen committed:
Update for changes to Command.params

ltzmax commented:
From testing branch `V3/dpy2_unstable` with the command `[p]queue` from audio:
```py
┌──────────────────────────────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────────────────────────────┐
│ C:\Users\maxbo\redenv\lib\site-packages\discord\ext\commands\core.py:182 in wrapped                                                                        │
│ >  182             ret = await coro(*args, **kwargs)                                                                                                       │
│ C:\Users\maxbo\redenv\lib\site-packages\redbot\cogs\audio\core\commands\queue.py:175 in command_queue                                                      │
│ > 175         return await menu(ctx, queue_page_list, queue_controls, page=(page - 1))                                                                     │
│ C:\Users\maxbo\redenv\lib\site-packages\redbot\core\utils\menus.py:118 in menu                                                                             │
│ > 109             if message.channel.permissions_for(ctx.me).manage_messages:                                                                              │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
AttributeError: 'PartialMessageable' object has no attribute 'permissions_for'
```
I am unsure how this happened and to why it happened. i only ran `queue` and waitied a bit then hit the `x` and the above error showed.

i had 132 songs in queue if that matter.

Jackenmen committed:
[threads] Support threads in ignore checks and allow ignoring them

- Updated `[p](un)ignore channel` to accept threads
- Updated `[p]ignore list` to list ignored threads
- Updated logic in `Red.ignored_channel_or_guild()`

Ignores for guild channels now work as follows (only changes for threads):
- if channel is not a thread:
    - check if user has manage channels perm in channel
      and allow command usage if so
    - check if channel is ignored and disallow command usage if so
    - allow command usage if none of the conditions above happened
- if channel is a thread:
    - check if user has manage channels perm in parent channel
      and allow command usage if so
    - check if parent channel is ignored and disallow command usage
      if so
    - check if user has manage thread perm in parent channel
      and allow command usage if so
    - check if thread is ignored and disallow command usage if so
    - allow command usage if none of the conditions above happened

Jackenmen committed:
[partial] Raise TypeError when channel is of PartialMessageable type

- Red.embed_requested
- Red.ignored_channel_or_guild

Jackenmen committed:
[partial] Discard command messages when channel is PartialMessageable

Jackenmen committed:
[threads] Add utilities for checking appropriate perms in both channels & threads

Jackenmen committed:
[threads] Update code to use can_react_in() and @bot_can_react()

Jackenmen committed:
[threads] Update code to use can_send_messages_in

Jackenmen committed:
[threads] Add send_messages_in_threads perm to mute role and overrides

Jackenmen committed:
[threads] Update code to use (bot/user)_can_manage_channel

Jackenmen committed:
[threads] Update [p]diagnoseissues to work with threads

Jackenmen committed:
Type hint fix

Jackenmen committed:
[threads] Patch vendored discord.ext.menus to check proper perms in threads

I guess we've reached time when we have to patch the lib we vendor...

Jackenmen committed:
Make docs generation work with non-final d.py releases

Jackenmen committed:
Update discord.utils.oauth_url() usage

Jackenmen committed:
Swap usage of discord.Embed.Empty/discord.embeds.EmptyEmbed to None

Jackenmen committed:
Update usage of Guild.member_count to work with `None`

Jackenmen committed:
Switch from Guild.vanity_invite() to Guild.vanity_url

Jackenmen committed:
Update startup process to work with d.py's new asynchronous startup

Jackenmen committed:
Use setup_hook() for pre-connect actions

Jackenmen committed:
Update core's add_cog, remove_cog, and load_extension methods

Jackenmen committed:
Update all setup functions to async and add awaits to bot.add_cog calls

Jackenmen committed:
Modernize cogs by using async cog_load and cog_unload

Jackenmen committed:
Address StoreChannel removal

Jackenmen committed:
[partial] Disallow passing PartialMessageable to Case.channel

Jackenmen committed:
[partial] Update cogs and utils to work better with PartialMessageable

- Ignore messages with PartialMessageable channel in CustomCommands cog
- In Filter cog, don't pass channel to modlog.create_case()
  if it's PartialMessageable
- In Trivia cog, only compare channel IDs
- Make `.utils.menus.menu()` work for messages
  with PartialMessageable channel
- Make checks in `.utils.tunnel.Tunnel.communicate()` more rigid

Jackenmen committed:
Add few missing DEP-WARNs

TrustyJAID reviewed (approved):
LGTM

Jackenmen commented:
YASSS
